fix: в MessageService.process() добавлена проверка на null перед выво…

Aleksandrinka opened the pull request:
…дом сообщения

В сигнатуру MessageService.process() добавлен порядок вывода - по возрастанию/по убыванию, порядковый номер сообщения добавляется к выводу с помощью OrderDecorator
Перегружен метод MessageService.process() добавлением в сигнатуру метода параметра обработки дублей - допустимы/недопустимы

feat: добавлен enum MessageOrder для хранения порядка вывода сообщений
добавлен enum Doubling для хранения состояния допустимости дублей в выводе
добавлен декоратор OrderDecorator, который добавляет к сообщению его порядковый номер
добавлен класс utils.MessageUtils для вспомогательной работы с массивами строк
добавлен метод MessageUtils.changeDoublesToNull() для изменения всех повторяющихся строк в массиве на null
добавлен метод MessageUtils.isMessageAlreadyInArray() для определения, входит ли строка в массив

Aleksandrinka committed:
fix: в MessageService.process() добавлена проверка на null перед выводом сообщения
В сигнатуру MessageService.process() добавлен порядок вывода - по возрастанию/по убыванию, порядковый номер сообщения добавляется к выводу с помощью OrderDecorator
Перегружен метод MessageService.process() добавлением в сигнатуру метода параметра обработки дублей - допустимы/недопустимы

feat: добавлен enum MessageOrder для хранения порядка вывода сообщений
добавлен enum Doubling для хранения состояния допустимости дублей в выводе
добавлен декоратор OrderDecorator, который добавляет к сообщению его порядковый номер
добавлен класс utils.MessageUtils для вспомогательной работы с массивами строк
добавлен метод MessageUtils.changeDoublesToNull() для изменения всех повторяющихся строк в массиве на null
добавлен метод MessageUtils.isMessageAlreadyInArray() для определения, входит ли строка в массив

AlexanderDolgorukov commented on src/main/java/com/tcs/edu/service/MessageService.java:
```diff
@@ -16,16 +20,83 @@ public class MessageService {
      * @param message  Обязательное сообщение
      * @param messages Набор сообщений для печати
      */
-    public static void process(Severity level, String message, String... messages) {
```
а старый метод совсем удалился?) его же хотели оставить, а новые просто добавлять перезагрузкой)

AlexanderDolgorukov commented on src/main/java/com/tcs/edu/service/MessageService.java:
```diff
@@ -16,16 +20,83 @@ public class MessageService {
+                    final String orderDecoratedMessage = OrderDecorator.decorate(currentMessage, currentMessageCount + 1);
+                    final String severityDecoratedMessage = SeverityDecorator.decorate(level, orderDecoratedMessage);
+                    final String timestampDecoratedMessage = TimestampMessageDecorator.decorate(severityDecoratedMessage);
+                    ConsolePrinter.print(timestampDecoratedMessage);
```
логика по подготовке к выводу timestampDecoratedMessage дублируется ниже - а это значит, что прямо просится выделение отдельного метода + рефакторинг)

Aleksandrinka replied:
Исправлено, добавлен метод MessageService.getAllDecorations()

AlexanderDolgorukov commented on src/main/java/com/tcs/edu/service/MessageService.java:
```diff
@@ -16,16 +20,83 @@ public class MessageService {
+                    final String severityDecoratedMessage = SeverityDecorator.decorate(level, orderDecoratedMessage);
+                    final String timestampDecoratedMessage = TimestampMessageDecorator.decorate(severityDecoratedMessage);
+                    currentMessageCount++;
+                    ConsolePrinter.print(timestampDecoratedMessage);
```
вот тут вот дублируется)

Aleksandrinka replied:
Исправлено

AlexanderDolgorukov commented on src/main/java/com/tcs/edu/service/MessageService.java:
```diff
@@ -16,16 +20,83 @@ public class MessageService {
+                        final String orderDecoratedMessage = OrderDecorator.decorate(currentMessage, currentMessageCount + 1);
+                        final String severityDecoratedMessage = SeverityDecorator.decorate(level, orderDecoratedMessage);
+                        final String timestampDecoratedMessage = TimestampMessageDecorator.decorate(severityDecoratedMessage);
+                        ConsolePrinter.print(timestampDecoratedMessage);
```
тут тоже как будто можно избавиться от дублирования))

Aleksandrinka replied:
Исправлено

Aleksandrinka committed:
feat: Убраны static методы и классы
Добавлен класс Message, содержащий тело сообщения и уровень важности
Рефакторинг

Aleksandrinka committed:
fix: Убрано дублирование кода, добавлен метод MessageService.getAllDecorations()

Aleksandrinka committed:
fix: Добавлены конструкторы в MessageService вместо явной инициализации полей

AlexanderDolgorukov commented on src/main/java/com/tcs/edu/service/MessageService.java:
```diff
@@ -12,20 +16,115 @@
+                Message currentMessage = messages[currentMessageCount];
+                if (currentMessage.getBody() != null) {
+                    final String decoratedMessage =
+                            getAllDecorations(currentMessage, currentMessageCount + 1);
```
можно ++ вместо +1 использовать, чтобы читаемость лучше была :)

AlexanderDolgorukov commented on src/main/java/com/tcs/edu/service/MessageService.java:
```diff
@@ -12,20 +16,115 @@
+            for (Message currentMessage : messages) {
+                if (currentMessage != null) {
+                    final String decoratedMessage =
+                            getAllDecorations(currentMessage, currentMessageCount);
```
на строке 60 ты передаешь currentMessageCount+1, а тут currentMessageCount и ниже делаешь currentMessageCount++ - а почему так?)

AlexanderDolgorukov commented on src/main/java/com/tcs/edu/service/MessageService.java:
```diff
@@ -12,20 +16,115 @@
+            } else {
+                Integer currentMessageCount = 1;
+                for (Message currentMessage : messages) {
+                    if (currentMessage.getBody() != null & !MessageUtils.isMessageAlreadyInArray(currentMessage.getBody(), alreadyPrintedMessages)) {
```
вот этот вот участок кода тоже дублируется аналогично со строками 97-...) если есть дублирование - значит можно вынести в отдельную функцию)

Aleksandrinka committed:
fix: Добавлены конструкторы в MessageService вместо явной инициализации полей

AlexanderDolgorukov commented on src/main/java/com/tcs/edu/domain/Message.java:
```diff
@@ -19,4 +21,21 @@ public Severity getSeverity() {
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
для читаемости лучше вместо "o" назвать хотя бы "obj" - будет понятно, что за переменная)

AlexanderDolgorukov commented on src/main/java/com/tcs/edu/Application.java:
```diff
@@ -27,5 +27,8 @@ public static void main(String[] args) {
 
+        System.out.println(new Message("Hello again2", Severity.MINOR));
+        System.out.println((new Message("Hello again2", Severity.MINOR)).equals(new Message("Hello again2", Severity.MINOR)));
+        System.out.println((new Message("Hello again2", Severity.MINOR)).equals(new Message("Hello again2", Severity.MAJOR)));
```
а с помощью метода hashCode() пробовала?)
